itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestStatsReplicationScenarios.java

```diff
@@ -301,12 +338,106 @@ private String dumpLoadVerify(List<String> tableNames, String lastReplicationId,
+        if (args.dbName.equalsIgnoreCase(replicatedDbName) && cntTables > failAfterNumTables) {
+          injectionPathCalled = true;
+          LOG.warn("Verifier - DB : " + args.dbName + " TABLE : " + args.tblName);
+          return false;
+        }
+        return true;
+      }
+    };
+
+    InjectableBehaviourObjectStore.setAlterTableModifier(callerVerifier);
+    try {
+      replica.loadFailure(replicatedDbName, tuple.dumpLocation);
+      callerVerifier.assertInjectionsPerformed(true, false);
+    } finally {
+      InjectableBehaviourObjectStore.resetAlterTableModifier();
+    }
+  }
+
+  private void failIncrementalLoad(WarehouseInstance.Tuple dumpTuple, int failAfterNumEvents) throws Throwable {
+    // fail add notification when updating table stats after given number of such events. Thus we
+    // test successful application as well as failed application of this event.
+    BehaviourInjection<NotificationEvent, Boolean> callerVerifier
+            = new BehaviourInjection<NotificationEvent, Boolean>() {
+      int cntEvents = 0;
+      @Override
+      public Boolean apply(NotificationEvent entry) {
+        cntEvents++;
+        if (entry.getEventType().equalsIgnoreCase(EventMessage.EventType.UPDATE_TABLE_COLUMN_STAT.toString()) &&
+            cntEvents > failAfterNumEvents) {
+          injectionPathCalled = true;
+          LOG.warn("Verifier - DB: " + entry.getDbName()
+                  + " Table: " + entry.getTableName()
+                  + " Event: " + entry.getEventType());
+          return false;
+        }
+        return true;
+      }
+    };
+
+    InjectableBehaviourObjectStore.setAddNotificationModifier(callerVerifier);
+    try {
+      replica.loadFailure(replicatedDbName, dumpTuple.dumpLocation);
+    } finally {
+      InjectableBehaviourObjectStore.resetAddNotificationModifier();
+    }
+    callerVerifier.assertInjectionsPerformed(true, false);
+
+    // fail add notification when updating partition stats for for the second time. Thus we test
+    // successful application as well as failed application of this event.
+    callerVerifier = new BehaviourInjection<NotificationEvent, Boolean>() {
+      int cntEvents = 1;
+
+      @Override
+      public Boolean apply(NotificationEvent entry) {
+        cntEvents++;
+        if (entry.getEventType().equalsIgnoreCase(EventMessage.EventType.UPDATE_PARTITION_COLUMN_STAT.toString()) &&
+            cntEvents > failAfterNumEvents) {
+          injectionPathCalled = true;
+          LOG.warn("Verifier - DB: " + entry.getDbName()
+                  + " Table: " + entry.getTableName()
+                  + " Event: " + entry.getEventType());
+          return false;
+        }
+        return true;
+      }
+    };
+
+    InjectableBehaviourObjectStore.setAddNotificationModifier(callerVerifier);
+    try {
+      replica.loadFailure(replicatedDbName, dumpTuple.dumpLocation);
+    } finally {
+      InjectableBehaviourObjectStore.resetAddNotificationModifier();
+    }
```

Comment:
Shall add validations to see if REPL LOAD fails at right place.

Reply:
Done.